build.sh

```diff
@@ -372,6 +396,50 @@ if (( ${NUMARGS} == 0 )) || hasArg libraft || hasArg docs || hasArg tests || has
         cmake --build  "${LIBRAFT_BUILD_DIR}" ${VERBOSE_FLAG} -j${PARALLEL_LEVEL} --target ${CMAKE_TARGET}
       fi
   fi
+  compile_end=$(date +%s)
+  compile_total=$(( compile_end - compile_start ))
+
+  if [[ "$BUILD_REPORT_METRICS" == "ON" && -f "${LIBRAFT_BUILD_DIR}/.ninja_log" ]]; then
+      if ! rapids-build-metrics-reporter.py 2> /dev/null && [ ! -f rapids-build-metrics-reporter.py ]; then
+          echo "Downloading rapids-build-metrics-reporter.py"
+          curl -sO https://raw.githubusercontent.com/rapidsai/build-metrics-reporter/v1/rapids-build-metrics-reporter.py
+      fi
+
+      echo "Formatting build metrics"
+      MSG=""
+      # get some sccache/ccache stats after the compile
+      if [[ "$BUILD_REPORT_INCL_CACHE_STATS" == "ON" ]]; then
+          if [[ ${CACHE_COMMAND} == "sccache" && -x "$(command -v sccache)" ]]; then
+              COMPILE_REQUESTS=$(sccache -s | grep "Compile requests \+ [0-9]\+$" | awk '{ print $NF }')
+              CACHE_HITS=$(sccache -s | grep "Cache hits \+ [0-9]\+$" | awk '{ print $NF }')
+              HIT_RATE=$(echo - | awk "{printf \"%.2f\n\", $CACHE_HITS / $COMPILE_REQUESTS * 100}")
+              MSG="${MSG}<br/>cache hit rate ${HIT_RATE} %"
+          elif [[ ${CACHE_COMMAND} == "ccache" && -x "$(command -v ccache)" ]]; then
+              CACHE_STATS_LINE=$(ccache -s | grep "Hits: \+ [0-9]\+ / [0-9]\+" | tail -n1)
+              if [[ ! -z "$CACHE_STATS_LINE" ]]; then
+                  CACHE_HITS=$(echo "$CACHE_STATS_LINE" - | awk '{ print $2 }')
+                  COMPILE_REQUESTS=$(echo "$CACHE_STATS_LINE" - | awk '{ print $4 }')
+                  HIT_RATE=$(echo - | awk "{printf \"%.2f\n\", $CACHE_HITS / $COMPILE_REQUESTS * 100}")
+                  MSG="${MSG}<br/>cache hit rate ${HIT_RATE} %"
+              fi
+          fi
+      fi
+      MSG="${MSG}<br/>parallel setting: $PARALLEL_LEVEL"
+      MSG="${MSG}<br/>parallel build time: $compile_total seconds"
+      if [[ -f "${LIBRAFT_BUILD_DIR}/libraft.so" ]]; then
+          LIBRAFT_FS=$(ls -lh ${LIBRAFT_BUILD_DIR}/libraft.so | awk '{print $5}')
+          MSG="${MSG}<br/>libraft.so size: $LIBRAFT_FS"
+      fi
```

Comment:
I'd like to include @vyasr and @davidwendt in a discussion here about how we can eliminate some of this boilerplate code.

Presumably some of this functionality will be needed by other repositories, so we should strive to make `rapids-build-metrics-reporter.py` handle as much of this as possible so that we have a single source of truth.

I have some thoughts and proposals here:

- I think `rapids-build-metrics-reporter.py` should handle `sccache -s`/`ccache -s` parsing. The lines of code in this PR are already duplicated from cudf's [build.sh](https://github.com/rapidsai/cudf/blob/4c4fdd2c86deef16bb68a13bc401193274c45266/build.sh#L305-L310), so this is a prime candidate for deduping. Perhaps we can expose a `--cache-tool`/`-c` flag that will let consumers indicate whether `ccache` or `sccache` should be used.
- Whenever possible, I'm in favor of avoiding the manual generation of HTML in bash variables (e.g. `<br/>`). In place of the `--msg` flag (which appears to take in an HTML string), I think we should create the following flags:
  - `--stat`/`-s` - this flag should be able to be specified multiple times and its value will be a key/value pair that can be added to a final summary table (e.g. `--stat="Parallel Setting=$PARALLEL_LEVEL"`, `--stat="Parallel Build Time=$compile_total seconds"`)
  - `--file-size`/`-f` - this flag should be able to be specified multiple times and its value will be a path to a file whose name and size will be added to a final summary table (e.g. `--file-size="${LIBRAFT_BUILD_DIR}/libraft.so"`)

The sum of these changes should allow us to dedupe quite a bit here and also eliminate the `MSG_OUTFILE` lines in this PR.

Once we get a consensus on all of this, I will open an issue to track the rollout of this new script to other repositories as well.

Reply:
I'd rather have the .py strictly handle the `.ninja_log` file. It currently supports other formats besides html including csv and xml (unittest format?). Both of these are handy at times and and there is no place for the stats in these outputs.
Rather than including the ccache stats in the html output, maybe we generate a separate file with sccache states into the artifacts directory. The reason is what incorporated in the html output originally was to just minimize the burden on the Jenkins job at the time. Just piping the `sccache -s > $RAPIDS_ARTIFACT_DIR/cache-stats.txt` seems a reasonable approach.

Reply:
So we entirely eliminate formating sccache/ccache statistics and just add it as a raw file for the user to parse?

Reply:
> Perhaps we can expose a --cache-tool/-c flag

This already exists in RAFT!

Reply:
If you are calling the output of `sccache -s` raw output then yes.

Reply:
> I'd rather have the .py strictly handle the `.ninja_log` file. It currently supports other formats besides html including csv and xml (unittest format?). Both of these are handy at times and and there is no place for the stats in these outputs.
> Rather than including the ccache stats in the html output, maybe we generate a separate file with sccache states into the artifacts directory. The reason is what incorporated in the html output originally was to just minimize the burden on the Jenkins job at the time. Just piping the `sccache -s > $RAPIDS_ARTIFACT_DIR/cache-stats.txt` seems a reasonable approach.

I see. I wasn't aware of these other formats.

I think there's value in having a single HTML file summary that includes all relevant build stats for CI as opposed to the stats being strewed across many files.

I wonder if we can make some changes to our Python script to accommodate both uses.

What if we rename `rapids-build-metrics-reporter.py` to `rapids-reporter.py` and implement subcommands for each report type?

For instance:

- `rapids-reporter.py summary` can produce a single HTML file and support all of the flags I mentioned above
- `rapids-reporter.py ninja` can support multiple output file types and support a separate set of flags

This should be doable with argparse. @davidwendt, what are your thoughts on this?

I realize this adds a bit of complexity, but I think the upfront investment is worth it. Plus we can write unit tests for the Python script to ensure it continues to work.

The end goal here is to strike the right balance between reusability, simplicity, and user experience for RAPIDS users/repos.

> > Perhaps we can expose a --cache-tool/-c flag
> 
> This already exists in RAFT!

I was referring to creating this option in the Python script. Presumably its value would be set to the value passed into `raft`'s `build.sh` script.

Reply:
I think centralizing more of the logic is the right approach as long as it is done in a way that the different pieces of the document (cache stats, ninja log, etc) are composable and extensible by packages that want to add new files etc. I would advocate for starting by building out the different pieces and then figuring out how to combine them afterwards. As long as there is a consistent HTML output the combining shouldn't be too difficult.

For building up the HTML in Python a reasonable low-level approach is building up the HTML tree with [xml.etree](https://docs.python.org/3/library/xml.etree.elementtree.html). A nicer approach might be to define a [Jinja template](https://jinja.palletsprojects.com/en/3.1.x/) that can be filled in, leaving some arbitrary extra sections that could be appended to so that the output format is extensible.

Reply:
> Are you suggesting that we just move the existing logic to another shell script? That doesn't seem to solve the problem of reusability / keeping things DRY. All of the repositories will just have another rapids-reporter.sh script that contains all of the duplicated lines of code still.

The existing shell script logic does not necessarily have move to into a new shell script. I just don't want to move this part of the existing shell script into the `rapids-build-metrics-reporter.py` file. We could move the shell script logic into a new separate .py file if that helps. I'm suggesting there will be 2 files in this repository -- one to build/format the sccache/ccache results and one to build/format the ninja-log results. And maybe a 3rd file to combine them if necessary.

> can you point out where / how the other output formats that you mentioned will be used? As far as I can tell, build.sh only outputs a single format type. Are you trying to preserve the multiple output formats for users who may run this tool manually outside of build.sh?

The other output formats are not being used by CI currently. I'm hoping the xml can be used in the future to fail a build if there is a problem with specific source file compile times. The CSV format has been helpful for me locally so I can load the results into Excel and sort them in different ways as well as compare them with other runs.

Reply:
I'd like to merge what we have in RAFT at this moment, and when there is consensus on how to share these scripts and how much logic to pluck out we can go ahead and update it. Is everyone okay with that? There are some major changes happening in RAFT right now where it would be very valuable to have these metrics for hindsight.